RFCs/0000-add-pdk-release.md

```diff
@@ -0,0 +1,118 @@
+
+- PDK subcommands should be reasonably extensible without any single command or action being overloaded with multiple,
+  distinct functions.
+
+## Proposed Implementation Design
+
+Implement a new pdk subcommand, `pdk release` which will offer the following actions.
+
+- `pdk release prep`
+
+  - Performs all the pre-release checks to ensure module is ready to be packaged, prompts user for missing info,
+    modifies files on disk as needed, etc.
+
+  - User is directed to review changes and commit the new state of the module’s code to their version control system.
+    (The exact contents of this prompt can be predicated on whether or not there is VCS metadata present in the module’s
+    working directory.)
+
+  - Ideally the user would be able to customize the pre-release checks that were performed, including defining their own
+    custom checks. This customization could be implemented through the [proposed `pdk config`
+    subcommand](https://github.com/puppetlabs/pdk-planning/pull/4).
+
+  - When run non-interactively (e.g. as part of a continuous delivery pipeline), it will exit non-zero if it would
+    require any user input to complete successfully.
+
+- `pdk release build [--target-dir=<value>]`
+
+  - Replicates the existing `pdk build` functionality with the modification that it will fail and exit non-zero if any
+    of the checks performed by `pdk release prep` would result in modification to the module’s code.
+
+  - If there is VCS metadata present in the module, this action could also fail and exit non-zero if there are changes
+    to the module’s code that have not been committed to VCS.
+
+  - If there is VCS metadata present in the module, this action could also automatically add a tag to the VCS repository
+    based on the module’s version string, or direct the user to do so.
+
+  - This command has no interactive elements, it either succeeds or fails.
+
+- `pdk release push [tarball] [--force]`
+
+  - _All functionality of this action is pending availability of proper Forge publish API._
+
+  - Publishes the given module tarball to the Forge.
+
+  - When not passed a path to a tarball, looks in `metadata.json` for the current version and then looks for a tarball
+    in the default build path matching that version and pushes it to Forge. Exits non-zero if it can’t find a tarball
+    matching the current metadata.json version.
+
+  - Defaults to asking for confirmation when run interactively, but can be disabled with the `--force` flag. Note that a
+    push will still fail, regardless of the `--force` flag, if a release of the module with the same version already
+    exists on the Forge.
+
+  - When run non-interactively (e.g. as part of a continuous delivery pipeline), it will default to not asking for
+    confirmation (`--force` behavior) and exit non-zero if it would require any user input to complete successfully.
+
+## Unresolved Questions
```

Comment:
I also have a question about whether this command would also look at the CI files templated by PDK. Or whether this is purely external to configuring the CI runs.

I put in a previous change so that TravisCI deployment step could easily be removed for those modules/sites that don't want publishing to the Forge (even accidentally).

Reply:
CI configs would have to be updated to explicitly call these new commands if the user wanted to publish to Forge. I doubt we would turn that on in the default templates but may expose customizability for that through `pdk config` down the line. (See also #4)

Reply:
The CI jobs definitely can/should be optional on forge credentials being configured. This will definitely require more manual intervention, e.g. to encrypt the secrets correctly for travis.